crates/core/src/live_socket/navigation/mod.rs

```diff
@@ -0,0 +1,85 @@
+mod ffi;
+
+use super::socket::LiveSocket;
+pub use ffi::*;
+use reqwest::Url;
+use std::sync::Arc;
+
+#[derive(Clone)]
+struct HandlerInternal(pub Option<Arc<dyn NavEventHandler>>);
+
+impl std::fmt::Debug for HandlerInternal {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if self.0.is_some() {
+            write!(f, "Handler Active")?;
+        } else {
+            write!(f, "No Handler Present")?;
+        };
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct NavCtx {
+    history: Vec<NavHistoryEntry>,
+    current_id: HistoryId,
+    navigation_event_handler: HandlerInternal,
+    current_dest: Option<NavHistoryEntry>,
+}
+
+impl NavHistoryEntry {
+    pub fn new(url: Url, id: HistoryId, state: Option<Vec<u8>>) -> Self {
+        Self {
+            url: url.to_string(),
+            id,
+            state,
+        }
+    }
+}
+
+impl NavCtx {
+    pub fn new() -> Self {
+        Self {
+            history: vec![],
+            current_id: 0,
+            navigation_event_handler: HandlerInternal(None),
+            current_dest: None,
+        }
```

Comment:
This looks more like an `impl Default for NavCtx` if you ask me.

Reply:
oh good catch, It was going to start with one item in it but I decided against it, the navigation should be fallibly wrapped by the socket.